### DISCUSSION TITLE
Set schema for Postgres connection outside of connection string

### DISCUSSION THREAD

**Discussion by amit-reflectivity** [prisma/prisma#5306] [Q&A] [answered]
I am able to work well with Prisma in my development environment as I can control the `DATABASE_URL` environment variable.

In the environments for staging and production that I am using there is an environment variable for the database connection string (`DATABASE_URL`) that can't be modified to add a schema.

I have tried to set the `search_path` for Postgres to only be the desired schema.

Prisma defaults to `public`. Other query tools no longer need to reference schema to access tables in that schema.

Is there any way to set the schema outside of the connection string?

If I could do something like this

```prisma
url = env("DATABASE_URL")?schema=alternative
```

OR

```prisma
url = env("DATABASE_URL")env("SCHEMA")
```

I tried to use an alternative environment variable

```prisma
url = env("SCHEMA")
```

```
(node:55846) UnhandledPromiseRejectionWarning: Error: error: Environment variable not found: DATABASE_URL.
--> schema.prisma:7
|
6 | provider = "postgresql"
7 | url = env("DATABASE_URL")
```

**Comment by nikolasburk** [comment, marked as answer]
Hey @amit-reflectivity 👋 

> Is there any way to set the schema outside of the connection string?

I don't think this is possible at the moment, but this would be an [interesting feature request that I'd encourage you to create](https://github.com/prisma/prisma/issues/new?assignees=&labels=&template=feature_request.md&title=) so that we can see if we can bring this functionality to Prisma!

In my understanding, using a different environment variable as you tried _should_ work though 🤔  did you run `prisma generate` after you changed the env var in the Prisma schema? It might be that Prisma Client still doesn't know about the new env var because it did not get re-generated.

Also, note that you can always override the connection URL from the Prisma schema in the `PrismaClient` constructor. HEre's an [example](https://www.prisma.io/docs/concepts/components/prisma-client/constructor#datasources) for how you can achieve this.

Let me know if that helps or if you have any further questions 🙌

**Reply by amit-reflectivity** [reply]
Hi @nikolasburk ,

Thanks for your response. On the different environment variable, I keep forgetting to run **prisma generate** (first day using Prisma) so my guess that is the problem. This was not a viable workaround as the DATABASE_URL can change and you can't declare compound environment variables.

After posting I tried other alternatives, in Postgres you can change the search_path that allows you change the schemas search when doing a SQL query. Before changing the search path I had to prefix the schema for the tables in the query. After changing the search path to be the desired schema then I no longer had to prefix for the tables in the query. From the error messages it seems given no schema setting that by default that Prisma will prefix **public** onto a table. If Prisma did not append **public** which already in the default search path then what I attempted would have solved my issue. Alternatively if there is a way to say not to append the schema but leave it to the search path. This would allow for using multiple schemas (may not be a good idea) in a single connection. In the system that I am doing there is a case to have two schemas since part of the data synced from an external source. It would keep the internal and external tables be separated by schema.

I thought about doing the override, I have setup PrismaService from following how to use with NestJS. i was thinking of doing it in a super() but wasn't sure how to proceed.

Would this work, in the class that is below?
```ts
super({
  datasources: {
   my_database: {
     url: 'file:./dev_qa.db'
   }
  },
}
```
```ts
import { Injectable, OnModuleInit, OnModuleDestroy } from '@nestjs/common';
import { PrismaClient } from '@prisma/client';

@Injectable()
export class PrismaService extends PrismaClient implements OnModuleInit, OnModuleDestroy {
  async onModuleInit() {
    await this.$connect();
  }
  async onModuleDestroy() {
    await this.$disconnect();
  }
}
```

**Reply by ryands17** [reply]
@amit-reflectivity Yes this will work fine 👍

**Reply by amit-reflectivity** [reply]
Hi @nikolasburk,

Added a feature request as suggested.
https://github.com/prisma/prisma/issues/5337

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
